lib/harmony/resources/storagemgr/storagemgt.go

```diff
@@ -0,0 +1,251 @@
+// Storagemgr is a utility for harmonytask tasks
+// to manage the available space.
+// Ex:
+//
+//	TaskTypeDefinition{
+//	   Storage: storMgr.MakeFuncs(UsageTemporary, 1<<30, func(taskID int) string {
+//					return fmt.Sprintf("task-%d", taskID)   // or use SQL to get the sector ID
+//			}),
+//	}
+//
+// Also useful: s.CombineFuncs(s1, s2, s3)
+package storagemgr
+
+import (
+	"bytes"
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	logging "github.com/ipfs/go-log/v2"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/lotus/lib/harmony/resources"
+	"github.com/filecoin-project/lotus/storage/sealer/storiface"
+)
+
+var log = logging.Logger("harmonytask")
+
+type path string
+
+var claimsMx sync.Mutex
+var claims = map[path][]consumption{}
+
+var purposes = map[Usage][]path{}
+
+type Usage int
+
+const (
+	UsagePermanent = iota
+	UsageTemporary
+	UsageCache
+	UsageStaging
+)
+
+type StorageMgr struct { // Functions are attached here to force users to call New()
+}
+
+func New(paths []storiface.StoragePath) *StorageMgr {
+	// Note: this may be called multiple times per process.
+	// populate purposes (global)
+	for _, p := range paths {
+		// @reviewer help this logic make sense
+		var t Usage = UsagePermanent
+		if p.CanSeal {
+			t = UsageStaging
+		}
+		if strings.Contains(p.LocalPath, "cache") {
+			t = UsageCache
+		}
+		if strings.Contains(p.LocalPath, "tmp") {
+			t = UsageTemporary
+		}
+		purposes[t] = append(purposes[t], path(p.LocalPath))
+	}
+
+	return &StorageMgr{}
+}
+
+// For lotus-provider (but not other deps users).
+// At start-up, remove (rm -rf) all the contents of all the claims.json files.
+func (s *StorageMgr) Cleanup() {
+	for _, paths := range purposes {
+		for _, path := range paths {
+			for _, u := range getJSON(path) {
+				_ = os.RemoveAll(u.Location)
+			}
+			_ = os.Remove(string(path) + "/claims.json")
+		}
+	}
+}
+
+func (s *StorageMgr) MakeFuncs(purpose Usage, need uint64, namer func(taskID int) string) resources.Storage {
+	return resources.Storage{
+		HasCapacity: func() bool { _, ok := s.hasCapacity(purpose, need); return ok },
+		Claim: func(id int) (string, error) {
+			name := namer(id)
+			return s.claim(purpose, need, name)
+		},
+		MarkComplete: s.markComplete,
+	}
+}
+
+type consumption struct {
+	Location string
+	Need     uint64
+}
+
+func getJSON(path path) []consumption {
+	b, err := os.ReadFile(string(path) + "/claims.json")
```

Comment:
* Should use filepath.Join (rest of the storage subsystem does pretty religously, and that will make our lives so much easier when we'll want to build super light storage-only SaaS backed thing for windows)

Reply:
Also I'd avoid putting anything in the scheduling codepath that makes any syscalls to the storage paths; Those will be backed by devices which are constantly hammered, and you will see basic things like statfs hang for 30 minutes

IMO reservations should live in the DB, addressed like the rest of the storage subsystem by existing pathIDs (or just use/interface with the existing storage subsystem which has reservation mechanism)